### PR TITLE
Add a homeDir field in LocalFileSystem for removefile to check

### LINKED ISSUE #4529: Feature: Check whether the path to delete is a subdirectory of kuzu database path
### API

Python

### Description

The current implementation of `removeFileIfExists(const std::string& path)` is unsafe since it can delete arbitrary files on disk.
We should consider adding a guard that it can only delete files under the current running database path.

### PR DESCRIPTION
**Pull request by SterlingT3485**
# Description

Added this for `removeFileIfExists(const std::string& path)` to only remove the files under our DBpath

Fixes #4529 

# Contributor agreement

- [x] I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

**Review by acquamarin — COMMENTED** [review]
Let's also add some tests.

File: src/main/database.cpp [hunk 1]
```diff
@@ -85,13 +87,25 @@ std::unique_ptr<storage::BufferManager> Database::initBufferManager(const Databa
         db.dbConfig.bufferPoolSize, db.dbConfig.maxDBSize, db.vfs.get(), db.dbConfig.readOnly);
 }
 
+std::string expandPath(main::ClientContext* context, const std::string& path) {
```
[the remark this answers is not in the document]
Comment (SterlingT3485) [reply src/main/database.cpp:90]: I think this case would be handled by the `isSubdirectory` function to convert to a abusolute path without .. and . 

I added the convert in the expandPath as well just in case

File: src/main/database.cpp [hunk 2]
```diff
@@ -85,13 +87,25 @@ std::unique_ptr<storage::BufferManager> Database::initBufferManager(const Databa
...
-    vfs = std::make_unique<VirtualFileSystem>();
     // To expand a path with home directory(~), we have to pass in a dummy clientContext which
     // handles the home directory expansion.
-    auto clientContext = ClientContext(this);
     const auto dbPathStr = std::string(dbPath);
-    databasePath = vfs->expandPath(&clientContext, dbPathStr);
+    auto clientContext = ClientContext(this);
+    databasePath = expandPath(&clientContext, dbPathStr);
+
+    vfs = std::make_unique<VirtualFileSystem>(databasePath);
```
[the remark this answers is not in the document]
Comment (SterlingT3485) [reply src/main/database.cpp:110]: moved to base

File: src/include/common/file_system/local_file_system.h [hunk 3]
```diff
@@ -64,6 +68,9 @@ class KUZU_API LocalFileSystem final : public FileSystem {
     void truncate(FileInfo& fileInfo, uint64_t size) const override;
 
     uint64_t getFileSize(const FileInfo& fileInfo) const override;
+
+private:
+    std::string homeDir;
```
[the remark this answers is not in the document]
Comment (SterlingT3485) [reply src/include/common/file_system/local_file_system.h]: moved

File: src/include/common/file_system/local_file_system.h [hunk 4]
```diff
@@ -27,6 +27,10 @@ struct LocalFileInfo : public FileInfo {
 
 class KUZU_API LocalFileSystem final : public FileSystem {
 public:
+    LocalFileSystem() : homeDir("") {}
```
[the remark this answers is not in the document]
Comment (SterlingT3485) [reply src/include/common/file_system/local_file_system.h]: removed

File: src/common/file_system/local_file_system.cpp [hunk 5]
```diff
@@ -247,9 +247,28 @@ void LocalFileSystem::createDir(const std::string& dir) const {
...
+    // Resolve paths to absolute and normalize them
+    auto basePath = std::filesystem::absolute(base).lexically_normal();
+    auto subPath = std::filesystem::absolute(sub).lexically_normal();
+
+    // Convert paths to generic strings for consistent comparison
+    auto baseStr = basePath.generic_string();
+    auto subStr = subPath.generic_string();
+
+    // Check if subStr starts with baseStr and ensure proper boundaries
+    return subStr.starts_with(baseStr) &&
```
[the remark this answers is not in the document]
Comment (SterlingT3485) [reply src/common/file_system/local_file_system.cpp]: Changed to a safer solution , which uses filesystem::relative to check the relation between two absolute dirs, should be safe enough

**Comment by SterlingT3485** [comment]
> Let's also add some tests.

Tests are added via c api

File: test/c_api/database_test.cpp [hunk 6]
```diff
@@ -100,3 +102,57 @@ TEST_F(CApiDatabaseTest, CreationHomeDir) {
...
+    std::filesystem::remove_all("/tmp/dbHome/test1");
+}
+
+TEST_F(CApiDatabaseTest, VirtualFileSystemDeleteFilesEdge) {
+    std::string homeDir = "/tmp/dbHome/";
+    kuzu::common::VirtualFileSystem vfs(homeDir);
+    std::filesystem::create_directories("/tmp/dbHome/../test2");
+    std::filesystem::create_directories("/tmp/dbHome/test2");
+
+    std::cout << "Resolved Path: " << std::filesystem::absolute("/tmp/dbHome/../test2")
```
Comment (acquamarin) [inline test/c_api/database_test.cpp:137]: remove the cout messages

File: src/common/file_system/local_file_system.cpp [hunk 7]
```diff
@@ -220,10 +220,10 @@ void LocalFileSystem::createDir(const std::string& dir) const {
             // LCOV_EXCL_STOP
         }
         auto directoryToCreate = dir;
-        if (directoryToCreate.ends_with('/')) {
+        if (directoryToCreate.ends_with('/') || directoryToCreate.ends_with('\\')) {
```
Comment (acquamarin) [inline src/common/file_system/local_file_system.cpp:223]: If the `\\` is only used for windows, i recommend put the prediction inside the windows macro `#ifdef _WIN32`

File: test/c_api/database_test.cpp [hunk 8]
```diff
@@ -100,3 +102,57 @@ TEST_F(CApiDatabaseTest, CreationHomeDir) {
...
+    }
+
+    vfs.removeFileIfExists("/tmp/dbHome/test2");
+
+    ASSERT_TRUE(std::filesystem::exists("/tmp/test2"));
+    ASSERT_FALSE(std::filesystem::exists("/tmp/dbHome/test2"));
+
+    // Cleanup: Remove directories after the test
+    std::filesystem::remove_all("/tmp/test2");
+    std::filesystem::remove_all("/tmp/dbHome/test2");
```
Comment (acquamarin) [inline test/c_api/database_test.cpp:157]: Can you also add those test cases:
1. Delete directories with edge cases.
2. Test whether it works as expected on windows since windows paths can have a mixture of '\\' and '/'.
For example: homedir:  "C:\Desktop\dir"
Path to delete: "C:\Desktop/dir/test1" 
3. Test whether we error in wild card pattern paths. (e.g. path to delete: '/tmp/kuzu*.test')
4. Test when we delete a path with homedirectory symbol (e.g. path to delete "~/test")

File: src/include/common/file_system/local_file_system.h [hunk 9]
```diff
@@ -27,6 +27,9 @@ struct LocalFileInfo : public FileInfo {
 
 class KUZU_API LocalFileSystem final : public FileSystem {
 public:
+    LocalFileSystem() = default;
+    explicit LocalFileSystem(std::string homeDir) : FileSystem(std::move(homeDir)) {}
```
[the remark this answers is not in the document]
Comment (SterlingT3485) [reply src/include/common/file_system/local_file_system.h:31]: I think it's because we are using creating dummy local file system just to use its member function in many place? We can change to pass in empty string as homeDir in such usages.

File: test/c_api/database_test.cpp (hunk 6 above)
Comment (SterlingT3485) [reply test/c_api/database_test.cpp:137]: removed

File: src/common/file_system/local_file_system.cpp (hunk 7 above)
Comment (SterlingT3485) [reply src/common/file_system/local_file_system.cpp:223]: wrapped this with macro

File: src/include/common/file_system/local_file_system.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (SterlingT3485) [reply src/include/common/file_system/local_file_system.h:30]: removed this default constructor

File: test/c_api/database_test.cpp (hunk 8 above)
Comment (SterlingT3485) [reply test/c_api/database_test.cpp:232]: add more tests, we do not support wild card patter for safety.

**Review by acquamarin — APPROVED** [review]
